byway/src/main/webapp/recommendations.js

```diff
@@ -0,0 +1,204 @@
+  const start = new google.maps.LatLng(37.7699298, -122.4469157);
+  const end = new google.maps.LatLng(37.7683909618184, -122.51089453697205);
+  const mapOptions = {
+    zoom: 14,
+    center: start
+  }
+  addDestinations(start, end);
+  map = new google.maps.Map(document.getElementById('map'), mapOptions);
+  directionsRenderer.setMap(map);
+  document.getElementById("route").addEventListener("click", () => {
+    calcRoute(directionsService, directionsRenderer, start, end);
+  });
+  placesService = new google.maps.places.PlacesService(map);
+}
+
+/**
+ * Add start and end points of the route to regions to search around.
+ * Temporary set up to mimic adding several endpoints between legs.
+ */
+function addDestinations(start, end) {
+  destinations.add(start);
+  destinations.add(end);
+  regions.push(...destinations);
+}
+
+/**
+ * Creates a route between two points and loads onto the map.
+ * Finds points along the path to load the regions Array.
+ * Loads recommendations centered around these points.
+ * @param {DirectionsService directionsService} finds directions
+ * @param {DirectionsRenderer directionsRenderer} renders the route
+ * @param {LatLng start} starting point location
+ * @param {LatLng end} ending point location
+ */
+function calcRoute(directionsService, directionsRenderer, start, end) {
+  const request = {
+      origin:  start,
+      destination: end,
+      travelMode: 'DRIVING'
+  };
+  directionsService.route(request, function(response, status) {
+    if (status == 'OK') {
+      directionsRenderer.setDirections(response);
+      findRegions(response);
+      loadRecommendations();
+    } else {
+      alert("Could not calculate route due to: " + status);
+    }
+  });
+}
+
+/**
+ * Goes through steps along route to find the average location
+ * between a step's start and end location. Store these points
+ * in the regions Array.
+ * @param {DirectionsResult directionResult} contains directions
+ * for the route made.
+ */
+function findRegions(directionResult) {
+  const myRoute = directionResult.routes[0].legs[0];
+  for(step of myRoute.steps) {
+    const avgLat = (step.start_location.lat() + step.end_location.lat()) / 2;
+    const avgLng = (step.start_location.lng() + step.end_location.lng()) / 2;
+    const avgLoc = {lat: avgLat, lng: avgLng};
+    if(step.distance.value > MIN_DISTANCE) {
+      regions.push(avgLoc);
+    }
+  }
+}
+
+/**
+ * Set a timeout to delay the browser.
+ * @param {Number delayMs} number of milliseconds
+ */
+function delayPromise(delayMs) {
+  return new Promise(resolve => setTimeout(resolve, delayMs));
+}
+
+/**
+ * Go through a user's interests and search for places
+ * fitting those interests. Search around the regions
+ * previously found using textSearch.
+ */
+async function loadRecommendations() {
+  let isLimited = false;
+  for(interest of interests) {
+    for(region of regions) {
+      await delayPromise(250);
```

Comment:
What's the purpose of this delay? Is it to prevent too many requests going to the maps API at the same time? Can you please add some comments explaining the logic here.

Reply:
Updated